h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -872,30 +626,75 @@ protected static SparseMatrixDimensions calculateCSRMatrixDimensions(Chunk[] chu
-        long rowIndicesCount = 0;
+    static SparseMatrixDimensions calculateCSRMatrixDimensions(Frame f, int[] chunkIds, Vec w, DataInfo di) {
+        CalculateCSRMatrixDimensionsMrFun fun = new CalculateCSRMatrixDimensionsMrFun(f, di, w, chunkIds);
+        H2O.submitTask(new LocalMR(fun, chunkIds.length)).join();
+        
+        return new SparseMatrixDimensions(ArrayUtils.sum(fun._nonZeroElementsCounts), ArrayUtils.sum(fun._rowIndicesCounts) + 1);
+    }
 
-        for (Integer chunk : chunks) {
-            for (long i = f.anyVec().espc()[chunk]; i < f.anyVec().espc()[chunk + 1]; i++) {
-                if (w != null && w.at(i) == 0) continue;
-                rowIndicesCount++;
+    private static class CalculateCSRMatrixDimensionsMrFun extends MrFun<CalculateCSRMatrixDimensionsMrFun> {
+        private Frame _f;
+        private DataInfo _di;
+        private Vec _w;
+        private int[] _chunkIds;
 
-                nonZeroElementsCount+= di._cats;
+        // OUT
+        private long[] _rowIndicesCounts;
+        private long[] _nonZeroElementsCounts;
 
-                for (int j = 0; j < di._nums; ++j) {
-                    double val = vecs[di._cats + j].at(i);
-                    if (val != 0) {
-                        nonZeroElementsCount++;
+        CalculateCSRMatrixDimensionsMrFun(Frame f, DataInfo di, Vec w, int[] chunkIds) {
+            _f = f;
+            _di = di;
+            _w = w;
+            _chunkIds = chunkIds;
+            _rowIndicesCounts = new long[chunkIds.length];
+            _nonZeroElementsCounts = new long[chunkIds.length];
+        }
+
+        @Override
+        protected void map(int i) {
+            final int cidx = _chunkIds[i];
+
+            long rowIndicesCount = 0;
+            long nonZeroElementsCount = 0;
+
+            if (_di._nums == 0) {
+                if (_w == null) {
+                    // no weights and only categoricals => sizing is trivial
+                    rowIndicesCount = _f.anyVec().chunkForChunkIdx(cidx)._len;
+                    nonZeroElementsCount = rowIndicesCount * _di._cats;
+                } else {
+                    Chunk ws = _w.chunkForChunkIdx(cidx);
+                    for (int r = 0; r < ws._len; r++)
+                        if (ws.atd(r) != 0) {
+                            rowIndicesCount++;
+                            nonZeroElementsCount += _di._cats; 
+                        }
+                }
+            } else {
+                Chunk[] cs = new Chunk[_di._nums];
+                for (int c = 0; c < cs.length; c++) {
+                    cs[c] = _f.vec(_di._cats + c).chunkForChunkIdx(cidx);
+                }
+                Chunk ws = _w != null ? _w.chunkForChunkIdx(cidx) : null;
+                for (int r = 0; r < cs[0]._len; r++) {
+                    if (ws != null && ws.atd(r) == 0)
+                        continue;
+                    rowIndicesCount++;
+                    nonZeroElementsCount += _di._cats;
```

Comment:
It could also be possible to do `nonZeroElementsCount += rowIndicesCount * _di._cats;` at the end to avoid addition per each row. Tiny speedup.

Reply:
Thanks, I will change it.